lib/yarp/ffi.rb

```diff
@@ -9,135 +12,200 @@ module YARP
-      type = type.strip.sub(/^const /, '')
-      type.end_with?('*') ? :pointer : type.to_sym
+      begin
+        raise unless yp_buffer_init(buffer)
+        yield buffer
+      ensure
+        yp_buffer_free(buffer)
+        buffer.pointer.free
+      end
     end
 
-    def self.load_exported_functions_from(header, functions)
-      File.readlines(File.expand_path("../../include/#{header}", __dir__)).each do |line|
-        if line.start_with?('YP_EXPORTED_FUNCTION ')
-          if functions.any? { |function| line.include?(function) }
-            /^YP_EXPORTED_FUNCTION (?<return_type>.+) (?<name>\w+)\((?<arg_types>.+)\);$/ =~ line or raise "Could not parse #{line}"
-            arg_types = arg_types.split(',')
-            arg_types = [] if arg_types == %w[void]
-            arg_types = arg_types.map { |type| resolve_type(type.sub(/\w+$/, '')) }
-            return_type = resolve_type return_type
-            attach_function name, arg_types, return_type
-          end
-        end
+    # This object represents a yp_string_t. We only use it as an opaque pointer,
+    # so it doesn't have to be an FFI::Struct.
+    class YPString
+      attr_reader :pointer
+
+      def initialize(pointer)
+        @pointer = pointer
       end
-    end
 
-    load_exported_functions_from("yarp.h",
-      %w[yp_version yp_parse_serialize yp_lex_serialize])
-    load_exported_functions_from("yarp/util/yp_buffer.h",
-      %w[yp_buffer_init yp_buffer_free])
-    load_exported_functions_from("yarp/util/yp_string.h",
-      %w[yp_string_mapped_init yp_string_free yp_string_source yp_string_length yp_string_sizeof])
+      def source
+        LibRubyParser.yp_string_source(pointer)
+      end
+
+      def length
+        LibRubyParser.yp_string_length(pointer)
+      end
+
+      def read
+        source.read_string(length)
+      end
+    end
 
+    # This is the size of a yp_string_t. It is returned by the yp_string_sizeof
+    # function which we call once to ensure we have sufficient space for the
+    # yp_string_t FFI pointer.
     SIZEOF_YP_STRING = yp_string_sizeof
 
-    def self.pointer(size)
-      pointer = FFI::MemoryPointer.new(size)
+    # Yields a yp_string_t pointer to the given block.
+    def self.with_string(filepath, &block)
+      string = FFI::MemoryPointer.new(SIZEOF_YP_STRING)
+
       begin
-        yield pointer
+        raise unless yp_string_mapped_init(string, filepath)
+        yield YPString.new(string)
       ensure
-        pointer.free
+        yp_string_free(string)
+        string.free
       end
     end
-
-    def self.yp_string(&block)
-      pointer(LibRubyParser::SIZEOF_YP_STRING, &block)
-    end
   end
+
+  # Mark the LibRubyParser module as private as it should only be called through
+  # the YARP module.
   private_constant :LibRubyParser
 
-  VERSION = LibRubyParser.yp_version.to_s
+  # The version constant is set by reading the result of calling yp_version.
+  VERSION = LibRubyParser.yp_version.read_string
```

Comment:
Self-note: actually we shouldn't get it from the lib but rather check it's the same and use templates or a `lib/yarp/version.rb` or so (from https://github.com/ruby/yarp/issues/765#issuecomment-1678717149)